File: backend/api_result.go
package backend

import "strings"

// ApiResult is the canonical backend response contract for frontend actions.
// It avoids fragile string-prefix parsing and provides stable error codes.
type ApiResult struct {
	OK      bool        `json:"ok"`
	Message string      `json:"message"`
	Code    string      `json:"code,omitempty"`
	Data    interface{} `json:"data,omitempty"`
}

func SuccessResult(message string) ApiResult {
	return ApiResult{OK: true, Message: message}
}

func SuccessResultWithData(message string, data interface{}) ApiResult {
	return ApiResult{OK: true, Message: message, Data: data}
}

func ErrorResult(code string, message string) ApiResult {
	return ApiResult{OK: false, Code: code, Message: message}
}

func legacyResult(result string) ApiResult {
	trimmed := strings.TrimSpace(result)
	lower := strings.ToLower(trimmed)

	if strings.HasPrefix(lower, "success") {
		return SuccessResult(trimmed)
	}
	if strings.HasPrefix(lower, "error") {
		return ErrorResult("LEGACY_ERROR", trimmed)
	}

	return SuccessResult(trimmed)
}

File: frontend/src/lib/api-result.js
export function parseApiResult(raw) {
  if (raw && typeof raw === 'object' && 'ok' in raw) {
    return {
      ok: Boolean(raw.ok),
      message:
        typeof raw.message === 'string' ? raw.message : raw.ok ? 'Success' : 'Request failed',
      code: typeof raw.code === 'string' ? raw.code : '',
      data: raw.data,
      raw,
    };
  }

  if (typeof raw === 'string') {
    const trimmed = raw.trim();
    const lower = trimmed.toLowerCase();
    if (lower.startsWith('success')) {
      return { ok: true, message: trimmed, code: '', data: undefined, raw };
    }
    if (lower.startsWith('error')) {
      return { ok: false, message: trimmed, code: 'LEGACY_ERROR', data: undefined, raw };
    }
    return { ok: true, message: trimmed || 'Success', code: '', data: undefined, raw };
  }

  if (raw === null || raw === undefined) {
    return {
      ok: false,
      message: 'No response received',
      code: 'NO_RESPONSE',
      data: undefined,
      raw,
    };
  }

  return { ok: true, message: String(raw), code: '', data: undefined, raw };
}
